src/net/client/load_balancer.rs

```diff
@@ -0,0 +1,1140 @@
+    /// Implementation of the new method.
+    fn new(receiver: mpsc::Receiver<ChanReq<Req>>) -> Self {
+        Self { receiver }
+    }
+
+    /// Run method.
+    pub async fn run(mut self) {
+        let mut next_id: u64 = 10;
+        let mut conn_stats: Vec<ConnStats> = Vec::new();
+        let mut conn_rt: Vec<ConnRT> = Vec::new();
+        let mut conns: Vec<Box<dyn SendRequest<Req> + Send + Sync>> =
+            Vec::new();
+
+        loop {
+            let req = match self.receiver.recv().await {
+                Some(req) => req,
+                None => break, // All references to connection objects are
+                               // dropped. Shutdown.
+            };
+            match req {
+                ChanReq::Add(add_req) => {
+                    let id = next_id;
+                    next_id += 1;
+                    conn_stats.push(ConnStats {
+                        _label: add_req.label,
+                        mean: (DEFAULT_RT_MS as f64) / 1000.,
+                        mean_sq: 0.,
+                        max_burst: add_req.max_burst,
+                        burst_interval: add_req.burst_interval,
+                        burst_start: Instant::now(),
+                        burst: 0,
+                        queue_length_plus_one: Arc::new(()),
+                    });
+                    conn_rt.push(ConnRT {
+                        id,
+                        est_rt: DEFAULT_RT,
+                        start: None,
+                        queue_length: 42, // To spot errors.
+                    });
+                    conns.push(add_req.conn);
+
+                    // Don't care if send fails
+                    let _ = add_req.tx.send(Ok(()));
+                }
+                ChanReq::GetRT(rt_req) => {
+                    let mut tmp_conn_rt = conn_rt.clone();
+
+                    // Remove entries that exceed the QPS limit. Loop
+                    // backward to efficiently remove them.
+                    for i in (0..tmp_conn_rt.len()).rev() {
+                        // Fill-in current queue length.
+                        tmp_conn_rt[i].queue_length = Arc::strong_count(
+                            &conn_stats[i].queue_length_plus_one,
+                        ) - 1;
+                        if let Some(max_burst) = conn_stats[i].max_burst {
+                            if conn_stats[i].burst_start.elapsed()
+                                > conn_stats[i].burst_interval
+                            {
+                                conn_stats[i].burst_start = Instant::now();
+                                conn_stats[i].burst = 0;
+                            }
+                            if conn_stats[i].burst > max_burst {
+                                tmp_conn_rt.swap_remove(i);
+                            }
+                        } else {
+                            // No limit.
+                        }
+                    }
```

Comment:
That would require two loops: one to update queue_length, burst_start and burst, and another for the filter. Do you think it is worth it?

Reply:
Updating those parameters can be done in forward order too, right?  Then you can do them using `.map()` or `.inspect()` in the same iterator chain.

Reply:
I'm confused. The current loop updates conn_stats. How would that work with map?

Reply:
Well, it only updates `conn_stats` at corresponding indices.  You can `.zip(&mut conn_stats)` or use `.enumerate()` to get the right index to use.

Reply:
Here's the full code:

```rust
let tmp_conn_rt: Vec<_> = conn_rt
    .iter()
    .cloned()
    .zip(&mut conn_stats)
    .filter_map(|(mut conn, stats)| {
        // Fill-in current queue length.
        conn.queue_length = Arc::strong_count(&stats.queue_length_plus_one) - 1;
        if let Some(max_burst) = stats.max_burst {
            if stats.burst_start.elapsed() > stats.burst_interval {
                stats.burst_start = Instant::now();
                stats.burst = 0;
            }
            if stats.burst > max_burst {
                // Filter out the connection.
                return None;
            }
        }
        Some(conn)
    })
    .collect();
```

Reply:
I don't see how this is more readable.

Reply:
It's more efficient and more idiomatic.  Readers don't need to consider how entries get filtered using `swap_remove()`, nor do they have to think about the backward iteration.  Instead of indexing `conn_stats`, we just naturally get the right element from it to work with.